deploy/web_scripts/notify_build.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+
+# Accept a web-hook from GitHub telling us about
+# a new built version of Hydrant.
+
+import json, requests
+
+from sys import stdin, stdout
+from os import environ, path
+from hmac import digest
+from tempfile import TemporaryDirectory
+from zipfile import ZipFile
+
+LOCKER_DIR = "/afs/sipb.mit.edu/project/hydrant/"
+
+OUTPUT_DIR = LOCKER_DIR + "web_scripts/hydrant/"
+
+CI_SECRETS_DIR = LOCKER_DIR + "ci_secrets/"
+HASH_SECRET = CI_SECRETS_DIR + "hash_secret"
+GITHUB_TOKEN = CI_SECRETS_DIR + "github_token"
+
+
+def main():
+    # Secret, used for HMAC input validation (so we know GitHub is being real)
+    with open(HASH_SECRET) as fh:
+        secret = fh.read().strip().encode("utf-8")
+    # API token for GitHub API requests (to get a path to the file)-- needs yearly updating!
+    with open(GITHUB_TOKEN) as fh:
+        token = fh.read().strip()
+
+    # Slurp content and validate with HMAC
+    body = stdin.read()
+    hexdigest = "sha256=" + digest(secret, body.encode("utf-8"), "sha256").hex()
+    if hexdigest != environ.get("HTTP_X_HUB_SIGNATURE_256", ""):
+        raise ValueError("bad digest")
+
+    # Extract the Run ID for the build
+    payload = json.loads(body)
+    if payload.get("action") != "completed":
+        raise ValueError("not completed")
+    job_id = payload.get("workflow_job", {}).get("run_id")
+    if not job_id:
+        raise ValueError("no job id")
+
+    # Fetch a list of artifacts from the GitHub API
+    response = requests.get(
+        "https://api.github.com/repos/sipb/hydrant/actions/runs/{}/artifacts".format(
+            job_id
+        )
+    )
+    if not response.ok:
+        raise ValueError("bad artifact fetch response: " + str(response.status_code))
+    artifact_info = response.json()
+
+    # For each known artifact:
+    success = False
+    for artifact in artifact_info.get("artifacts", []):
+        # check that its name is correct,
+        if artifact.get("name") != "built-site":
+            continue
+        url = artifact.get("archive_download_url")
+        if not url:
+            continue
+        # then fetch it.
+        response = requests.get(url, headers={"Authorization": ("Bearer " + token)})
+        with TemporaryDirectory() as td:
+            fname = path.join(td, "build_artifact.zip")
+            with open(fname, "wb") as fh:
+                for chunk in response.iter_content(chunk_size=4096):
+                    fh.write(chunk)
+            # Extract into the output directory.
+            with ZipFile(fname, "r") as zfh:
+                zfh.extractall(OUTPUT_DIR)
```

Comment:
does this overwrite silently? that's my main concern

in particular, due to hashing, it means that the folder will accumulate old build artifacts. does it need to be cleaned regularly then?

Reply:
I believe TemporaryDirectory() instantiates a dir in `/tmp` by default, which is a tmpfs and cleaned on reboot; as long as the machine is rebooted regularly, this shouldn't be an issue

Reply:
hm no, what i meant is in `OUTPUT_DIR`. note that the js files we serve have hashes (e.g. https://hydrant.mit.edu/static/js/main.ab9eab41.js)

Reply:
anyway, probably the best idea is to test this; do you / someone else want to uncomment it and try it?

Reply:
okay i tested this. yes, the old hashes are kept, so it might need cleaning every once in a while. we can figure this out later